main/utils.zkasm

```diff
@@ -955,4 +962,178 @@ maskAddress:
+    ; result is stored in: arithRes1(E) and arithOverflow(D)
+                    :CALL(mulARITH)
+    C => A
+    ; Check if modulus is 0 or 1
+    2 => B 
+    $               :LT, JMPC(zeroOneMod)
+    ; Now we will try to perform the following equation:
+    ; (k.l + k.h * 2²⁵⁶) * N + mulModResult = (D1 + D2) * 2²⁵⁶ + E 
+
+    ${(_mulMod / C) >> 256} => B     ; k.h
+    ; We can jump with Js, because later it's all verified by the ARITH
+    ${cond(B == 0)}  :JMPN(mulModNoKH) 
+
+    ; Since there's k.h we will split the equation in those 2
+    ;k.l * N + mulModResult = D1 * 2²⁵⁶ + E 
+    ;k.h * 2²⁵⁶ * N  = D2 * 2²⁵⁶ --> k.h  * N  = D2 
+
+    ; k.h  * N  = D2 
+    ; B * A + 0 = 0 * 2²⁵⁶ + E
+    ; D2 must be less than 2²⁵⁶
+    C => A ; Modulus
+    0 => C, D
+    ${B * A} => E   :MSTORE(tmpVarD), ARITH ; D2
+
+    ; k.l * N + mulModResult = D1 * 2²⁵⁶ + E 
+    ; B * A + C = D*2^256 + E
+    ; remember that:
+    ; result of mul is stored in: arithRes1(E) and arithOverflow(D)
+
+    ${(_mulMod / A) % (2 << 256)} => B   ; k.l
+    ${_mulMod % A} => C        ; mulModResult
+    ${(B * A + C) >> 256} => D ; D1
+    $               :MLOAD(arithRes1), ARITH
+
+    ; Finally we need to assert the following:
+    ; N>resultModulus
+    ; D1 + D2 = D
+    ; N>resultModulus   ; LT; ASSERT
+    A => B          ; modulus
+    C => A          ; mulModResult
+    $ => A          :LT
+    1               :ASSERT
+
+    ; Assert D1 + D2 = D ; ADD ;ASSERT
+    D => A ; D1
+    $ => B          :MLOAD(tmpVarD) ;D2
+    $ => A          :ADD
+    $               :MLOAD(arithOverflow), ASSERT
+                    :JMP(utilMULMODend)
+
+mulModNoKH:
+    ; if theres no K.h the equation is simplified as:
+    ; K * N + mulModResult = D*2^256 + E 
+    ; B * A + C = D*2^256 + E
+
+    C => A ; Modulus on A
+    ${(_mulMod / A)} => B   ; k
+    ${_mulMod % A} => C     ; mulModResult
+    $ => D         :MLOAD(arithOverflow)
+    $              :MLOAD(arithRes1), ARITH ; Can this be on the same line of MLOAD?
+
+    A => B          ; modulus
+    C => A          ; mulModResult
+    $ => A          :LT
+    1               :ASSERT
+                    :JMP(utilMULMODend)
+
+zeroOneMod:
+    0 => C
+
+utilMULMODend:
+    $ => A          :MLOAD(tmpVarA)
+    $ => B          :MLOAD(tmpVarB)
+    $ => D          :MLOAD(tmpVarD)
+    $ => E          :MLOAD(tmpVarE)
+    $ => RR         :MLOAD(tmpZkPC4)
+                    :JMP(RR)
+
+;@info exp(A,D) --> A^D
+;@in A, D => A^D
+;@out A => result
+
+expAD:
+    RR              :MSTORE(tmpZkPC3)
+    B               :MSTORE(tmpVarB3)
+    C               :MSTORE(tmpVarC3)
+    E               :MSTORE(tmpVarE3)
+    A => E
+    D => B
+    0 => A
+    $               :EQ,JMPC(expAD0)
+    E => A
+    0 => B
+    $               :EQ,JMPC(expA0D)
+    1 => C
+
+expADloop:
+
+    %MAX_CNT_ARITH - CNT_ARITH - 3 :JMPN(outOfCountersArith)
+
+    D => A
+    0 => B
+    $               :EQ,JMPC(expADend)
```

Comment:
Could make this check with Js, and then use ASSERT in the last interation to check that actually is 0, this way dont use a binary for loop